062-UseKRaft-feature-gate-promotion.md

```diff
@@ -0,0 +1,94 @@
+* JBOD support
+  * Tracked in [strimzi/strimzi-kafka-operator#9437](https://github.com/strimzi/strimzi-kafka-operator/issues/9437)
+  * Expected to be implemented in Apache Kafka 3.7.0.
+    The Strimzi implementation fill follow the release of Kafka 3.7.0 in Strimzi 0.40 or 0.41.
+
+## Proposed timeline
+
+This proposal proposes the following timeline for the graduation of the `UseKRaft` feature gate:
+* Move to _beta_ phase and be enabled by default in Strimzi 0.40.0
+* Move to _GA_ phase and be permanently enabled in Strimzi 0.42.0
+
+It is worth noting that:
+* In addition to the `UseKRaft` feature gate, the KRaft clusters are enabled / disabled using the `strimzi.io/kraft` annotation.
+  As a result, moving the `UseKRaft` feature gate to _beta_ or _GA_ does not mean that all new Kafka clusters have to use KRaft or that this has any impact on existing ZooKeeper based clusters.
+  KRaft will be applied only to the Kafka clusters with the right annotation.
+* Moving the `UseKRaft` feature to beta or GA does not mean we will drop support for ZooKeeper-based clusters.
+  While moving the `UseKRaft` feature gate to _GA_ defines the earliest moment when support for ZooKeeper based clusters can be dropped, it is currently not expected to happen right after Strimzi 0.42 and this proposal does not establish any such plan.
+
+Moving the `UseKRaft` feature gate to _beta_ or _GA_ serves mainly the following purpose:
+* Indicate the progress in Kraft implementation and improved production-readiness.
+* Make it easier to run KRaft clusters without the need to enable any feature gate.
+
+## Proposed changes
+
+Apart from the change to the feature gate status itself, this proposal also includes several other changes.
+
+### CRD changes
+
+KRaft is usable only with the use of `KafkaNodePool` custom resources that configure the number of replicas and storage configuration.
+As part of promoting the `UseKRaft` feature gate to _beta_ stage, we will make the following fields in the `Kafka` custom resource optional instead of required:
+* `.spec.kafka.replicas`
+* `.spec.kafka.storage`
+
+In addition, the KRaft clusters do not need the ZooKeeper configuration.
+So the `.spec.zookeeper` section will be made optional as well.
+
+For ZooKeeper clusters, the validation of these fields will be done inside the Strimzi operator.
+Additionally, CEL validation rules will be considered as they might allow us to do additional validation on the Kubernetes level.
+
+For Kafka clusters using the node pools, a warning will be raised by the operator when the ignored `.spec.kafka.replicas` and `.spec.kafka.storage` fields are used.
+For KRaft based clusters, a warning will be raised when `.spec.zookeeper` section is used.
+
+### Safety check for existing clusters
+
+For existing clusters, a safety check will be implemented to prevent users from switching existing ZooKeeper-based clusters to Kraft-based cluster (or vice-versa) by mistake.
```

Comment:
What kind of a safety check? Can you elaborate a little bit more please?

Reply:
I expanded this a bit to better explain what the check is. Essentially, the check should compare the desired metadata state (basically the KRaft annotation) with the actual metadata state (`.status.metadataState`) and if they do not match, it should reject it as the only way to transition between Zoo and KRaft should be the migration and we do not want to break ay cluster just because someone adds a wrong annotation to it.

Reply:
But this field doesn't not exist yet and it's part of my migration work (it's also called `kafkaMetadataState` as we agreed to call it after having `kafkaMetadataVersion` as well). If you need this field now for this proposal, I would like to try to get it from my PR also having the `KafkaMetadataStateManager` handling it in order to avoid to have a totally different way now and then my code for the migration. It would avoid a lot of conflicts for me maybe breaking everything else or anyway rewriting part of the code from scratch.

Reply:
> But this field doesn't not exist yet and it's part of my migration work (it's also called kafkaMetadataState as we agreed to call it after having kafkaMetadataVersion as well).

The migration proposal talks about `.status.metadataState`. I will update it at least in this proposal to use `kafkaMetadataState` then.

>If you need this field now for this proposal, I would like to try to get it from my PR also having the KafkaMetadataStateManager handling it in order to avoid to have a totally different way now and then my code for the migration. It would avoid a lot of conflicts for me maybe breaking everything else or anyway rewriting part of the code from scratch.

As this proposal says - I expect this to be either delivered by the migration proposal or separately. If you want, you can separate this from your Pr and we can look into it earlier. I think I already suggested it in the past. But I do not think this really needs to be covered in the proposal itself.

Reply:
> The migration proposal talks about .status.metadataState. I will update it at least in this proposal to use kafkaMetadataState then.

Yeah. There is a lot I should update on that proposal at some point (even the FSM is changed). We agreed to move from `metadataState` to 'kafkaMetadataState` when you introduced the `kafkaMetadataVersion` which was better than just `metadataVersion`.

> As this proposal says - I expect this to be either delivered by the migration proposal or separately. If you want, you can separate this from your Pr and we can look into it earlier. I think I already suggested it in the past. But I do not think this really needs to be covered in the proposal itself.

Yeah I know you suggested it in the past, but I thought about it like a way to breakdown the migration work while now it seems already useful (or needed?) to the `useKRaft` promotion proposal.

Reply:
Well, strictly speaking I think the current plan is to still deliver the migration in 0.40. So it might not be needed even without a separate PR. (but I'm fine if you want to do it earlier rather than wait of course)

Reply:
Well this is my concern, plans are done to be screwed up :-)